tracer/src/Datadog.Trace/OTelMetrics/MetricState.cs

```diff
@@ -91,39 +97,51 @@ public void RecordMeasurementDouble(double value, ReadOnlySpan<KeyValuePair<stri
         }
     }
 
-    public void RecordMeasurementGaugeLong(long value, ReadOnlySpan<KeyValuePair<string, object?>> tags)
-        => GetOrCreatePoint(tags).UpdateGauge(value);
-
-    public void RecordMeasurementGaugeDouble(double value, ReadOnlySpan<KeyValuePair<string, object?>> tags)
-        => GetOrCreatePoint(tags).UpdateGauge(value);
-
-    public void RecordMeasurementHistogramLong(long value, ReadOnlySpan<KeyValuePair<string, object?>> tags)
-        => GetOrCreatePoint(tags).UpdateHistogram(value);
-
-    public void RecordMeasurementHistogramDouble(double value, ReadOnlySpan<KeyValuePair<string, object?>> tags)
-        => GetOrCreatePoint(tags).UpdateHistogram(value);
-
-    public IEnumerable<MetricPoint> GetMetricPoints() => _points.Values;
-
-    public MetricStreamIdentity GetIdentity() => _identity;
+    /// <summary>
+    /// Builds metric point snapshots and adds them to the provided list.
+    /// For Delta temporality, resets the running values after snapshotting.
+    /// </summary>
+    public void TryBuildPoints(List<MetricPoint> into)
+    {
+        foreach (var point in _points.Values)
+        {
+            var snapshot = point.CreateSnapshotAndReset();
+            into.Add(snapshot);
+        }
+    }
 
     private MetricPoint GetOrCreatePoint(ReadOnlySpan<KeyValuePair<string, object?>> tags)
     {
         var tagSet = TagSet.FromSpan(tags);
 
-        var tagDict = new Dictionary<string, object?>();
-        for (int i = 0; i < tags.Length; i++)
+        if (_points.TryGetValue(tagSet, out var existingPoint))
         {
-            var kv = tags[i];
-            tagDict[kv.Key] = kv.Value;
+            return existingPoint;
         }
 
-        return _points.GetOrAdd(tagSet, _ => new MetricPoint(
-            _identity.InstrumentName,
-            _identity.MeterName,
-            _identity.InstrumentType,
-            MetricReaderHandler.GetTemporality(_identity.InstrumentType),
-            tagDict));
+        var tagsArray = tags.ToArray();  // Materialize span to heap array
```

Comment:
You should be able to remove `tagsArray` entirely and instead iterate over `tags` inside the callback